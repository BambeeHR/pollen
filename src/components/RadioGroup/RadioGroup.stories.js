import { action } from '@storybook/addon-actions';
import Form from '../../constants/Form';
import RadioGroup from './RadioGroup.vue';
import TypeOverline from '../TypeOverline/TypeOverline.vue';

const { Sizes, Variants } = Form;

const sizes = [Sizes.DENSE, Sizes.NORMAL];

export default {
  title: 'Toggles/RadioGroup',
  component: RadioGroup,
  parameters: {
    componentSubtitle: 'A group of `RadioInput` components',
  },
};

export const Gallery = () => ({
  components: { RadioGroup, TypeOverline },
  data() {
    return {
      sizes,
      options: [
        'Option 1',
        {
          label: 'My Second Option',
          value: 'foo',
        },
        {
          label: 'Option 3',
          value: 'bar',
        },
      ],
      value: 'foo',
      variants: [Variants.STANDARD, Variants.RAISED],
    };
  },
  methods: {
    handleInput: action('input'),
  },
  template: `
  <div class="flex -mx-8">
    <div class="flex-grow mx-8" v-for="variant in variants" :key="variant">
      <TypeOverline tag="h2" variant="large" class="mb-2">{{ variant }}</TypeOverline>
      <div v-for="size in sizes" :key="variant + '-' + size">
        <TypeOverline tag="h3" class="mb-2">{{ size }}</TypeOverline>
        <RadioGroup v-model="value" @input="handleInput" :variant="variant" :size="size" :options="options" />
        <TypeOverline tag="h3" class="mb-2">{{ size }}, invalid</TypeOverline>
        <RadioGroup v-model="value" @input="handleInput" :variant="variant" :size="size" :options="options" invalid />
        <TypeOverline tag="h3" class="mb-2">{{ size }}, disabled</TypeOverline>
        <RadioGroup v-model="value" @input="handleInput" :variant="variant" :size="size" :options="options" disabled />
      </div>
    </div>
  </div>
  `,
});

export const LabelOnTheLeft = () => ({
  components: { RadioGroup, TypeOverline },
  data() {
    return {
      sizes,
      options: [
        'Option 1',
        {
          label: 'My Second Option',
          value: 'foo',
        },
        {
          label: 'Option 3',
          value: 'bar',
        },
      ],
      value: 'foo',
      variants: [Variants.STANDARD, Variants.RAISED],
    };
  },
  methods: {
    handleInput: action('input'),
  },
  template: `
  <div class="flex -mx-8">
    <div class="flex-grow mx-8" v-for="variant in variants" :key="variant">
      <TypeOverline tag="h2" variant="large" class="mb-2">{{ variant }}</TypeOverline>
      <div v-for="size in sizes" :key="variant + '-' + size">
        <TypeOverline tag="h3" class="mb-2">{{ size }}</TypeOverline>
        <RadioGroup v-model="value" @input="handleInput" :variant="variant" :size="size" :options="options" label-position="left" />
        <TypeOverline tag="h3" class="mb-2">{{ size }}, invalid</TypeOverline>
        <RadioGroup v-model="value" @input="handleInput" :variant="variant" :size="size" :options="options" label-position="left" invalid />
        <TypeOverline tag="h3" class="mb-2">{{ size }}, disabled</TypeOverline>
        <RadioGroup v-model="value" @input="handleInput" :variant="variant" :size="size" :options="options" label-position="left" disabled />
      </div>
    </div>
  </div>
  `,
});

export const Vertical = () => ({
  components: { RadioGroup, TypeOverline },
  data() {
    return {
      sizes,
      options: [
        'Option 1',
        {
          label: 'My Second Option',
          value: 'foo',
        },
        {
          label: 'Option 3',
          value: 'bar',
        },
      ],
      value: 'foo',
      variants: [Variants.STANDARD, Variants.RAISED],
    };
  },
  methods: {
    handleInput: action('input'),
  },
  template: `
  <div class="flex -mx-8">
    <div class="flex-grow mx-8" v-for="variant in variants" :key="variant">
      <TypeOverline tag="h2" variant="large" class="mb-2">{{ variant }}</TypeOverline>
      <div v-for="size in sizes" :key="variant + '-' + size">
        <TypeOverline tag="h3" class="mb-2">{{ size }}</TypeOverline>
        <RadioGroup v-model="value" @input="handleInput" :variant="variant" :size="size" :options="options" vertical />
        <TypeOverline tag="h3" class="mb-2">{{ size }}, invalid</TypeOverline>
        <RadioGroup v-model="value" @input="handleInput" :variant="variant" :size="size" :options="options" invalid vertical />
        <TypeOverline tag="h3" class="mb-2">{{ size }}, disabled</TypeOverline>
        <RadioGroup v-model="value" @input="handleInput" :variant="variant" :size="size" :options="options" disabled vertical />
      </div>
    </div>
  </div>
  `,
});

export const VerticalWithLabelOnTheLeft = () => ({
  components: { RadioGroup, TypeOverline },
  data() {
    return {
      sizes,
      options: [
        'Option 1',
        {
          label: 'My Second Option',
          value: 'foo',
        },
        {
          label: 'Option 3',
          value: 'bar',
        },
      ],
      value: 'foo',
      variants: [Variants.STANDARD, Variants.RAISED],
    };
  },
  methods: {
    handleInput: action('input'),
  },
  template: `
  <div class="flex -mx-8">
    <div class="flex-grow mx-8" v-for="variant in variants" :key="variant">
      <TypeOverline tag="h2" variant="large" class="mb-2">{{ variant }}</TypeOverline>
      <div v-for="size in sizes" :key="variant + '-' + size">
        <TypeOverline tag="h3" class="mb-2">{{ size }}</TypeOverline>
        <RadioGroup v-model="value" @input="handleInput" :variant="variant" :size="size" :options="options" label-position="left" vertical />
        <TypeOverline tag="h3" class="mb-2">{{ size }}, invalid</TypeOverline>
        <RadioGroup v-model="value" @input="handleInput" :variant="variant" :size="size" :options="options" label-position="left" invalid vertical />
        <TypeOverline tag="h3" class="mb-2">{{ size }}, disabled</TypeOverline>
        <RadioGroup v-model="value" @input="handleInput" :variant="variant" :size="size" :options="options" label-position="left" disabled vertical />
      </div>
    </div>
  </div>
  `,
});

/* eslint-disable no-param-reassign */
[Gallery, LabelOnTheLeft, Vertical, VerticalWithLabelOnTheLeft].forEach(
  (item) => {
    item.story = {
      parameters: {
        jest: ['RadioGroup.test.js'],
      },
    };
  }
);
