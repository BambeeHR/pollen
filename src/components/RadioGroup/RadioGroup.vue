<template>
  <ToggleGroup
    class="radio-group"
    :options="options"
    :disabled="disabled"
    :invalid="invalid"
    :label-position="labelPosition"
    :name="name"
    :required="required"
    :size="size"
    :stacked="stacked"
    :value="value"
    :variant="variant"
    type="radio"
    @input="$emit('input', $event)"
  />
</template>

<script>
import Form from '../../constants/Form';
import ToggleGroup from '../internal/forms/toggles/ToggleGroup.vue';

/**
 * `import { RadioGroup } from '@bambee/pollen';`
 *
 * A component for selecting a single `value` from a group `options`. The
 * individual `RadioInput` component is available for more complex uses where,
 * for example, the label of the input needs to be a component rather than an
 * HTML string. In general howeer, this should be sufficient for most needs.
 */
export default {
  components: { ToggleGroup },
  props: {
    /** If the input is disabled. */
    disabled: {
      type: Boolean,
      default: false,
    },
    /**
     * If true, this field will display in an error state.
     */
    invalid: {
      type: Boolean,
      default: false,
    },
    /** Whether labels should display to the `left` or `right` of the radio. */
    labelPosition: {
      type: String,
      default: Form.LabelPositions.RIGHT,
      validator: (value) => Object.values(Form.LabelPositions).includes(value),
    },
    /**
     * The shared name between radio inputs. A random name is generated by
     * default.
     */
    name: {
      type: String,
      default: undefined,
    },
    /**
     * The available options for this component. This is an array of values.
     * Each individual value can be either a string, or an object with a `label`
     * and `value` key. The `label` is what the user sees. The `value` is what
     * will be emitted as an `input` event.
     */
    options: {
      type: Array,
      default: () => [],
    },
    /** If the input is disabled. */
    required: {
      type: Boolean,
      default: false,
    },
    /** One of `dense`, or `normal`. */
    size: {
      type: String,
      default: Form.Sizes.NORMAL,
      validator: (value) =>
        [Form.Sizes.DENSE, Form.Sizes.NORMAL].includes(value),
    },
    /** If true, displays radio options vertically. */
    stacked: {
      type: Boolean,
      default: false,
    },
    /**
     * Value of the input. Compatible with `v-model`.
     */
    value: {
      type: String,
      default: null,
    },
    /**
     * One of `standard` or `raised`.
     */
    variant: {
      type: String,
      default: Form.Variants.STANDARD,
      validator: (value) =>
        [Form.Variants.STANDARD, Form.Variants.RAISED].includes(value),
    },
  },
};
</script>
