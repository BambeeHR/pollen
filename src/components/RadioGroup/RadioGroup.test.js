import { mount } from '@vue/test-utils';
import RadioGroup from './RadioGroup.vue';

const options = [
  'Option 1',
  {
    label: 'Option 2',
    value: 'foo',
  },
  {
    label: 'Option 3',
    value: 'bar',
  },
];

describe('RadioGroup', () => {
  test('renders correctly', () => {
    const wrapper = mount(RadioGroup, {
      propsData: {
        name: 'test',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('renders options correctly', () => {
    const wrapper = mount(RadioGroup, {
      propsData: {
        options,
        value: 'foo',
        name: 'test',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('renders options correctly', () => {
    const wrapper = mount(RadioGroup, {
      propsData: {
        options,
        value: 'foo',
        name: 'test',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('emits an input event', async () => {
    const wrapper = mount(RadioGroup, {
      propsData: {
        options,
        value: null,
      },
    });
    const inputs = wrapper.findAll('input');
    expect(inputs.length).toBe(3);
    await inputs.at(1).trigger('click');
    await inputs.at(0).trigger('click');
    await inputs.at(2).trigger('click');
    expect(wrapper.emitted().input).toEqual([['foo'], ['Option 1'], ['bar']]);
  });
});
