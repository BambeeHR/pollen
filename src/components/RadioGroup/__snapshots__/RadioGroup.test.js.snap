// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup renders correctly 1`] = `
<div
  class="toggle-group radio-group toggle-group--horizontal toggle-group--normal"
>
  <label
    class="toggle-input radio-input toggle-input--normal toggle-input--standard toggle-input--label-right"
    for="test-0"
  >
    <input
      class="toggle-input__input"
      id="test-0"
      name="test"
      type="radio"
      value="Option 1"
    />
     
    <div
      class="radio-indicator radio-indicator--normal radio-indicator--standard"
    />
     
    <div
      class="toggle-input__label"
    >
      <span>
        Option 1
      </span>
    </div>
  </label>
  <label
    class="toggle-input radio-input toggle-input--normal toggle-input--standard toggle-input--label-right"
    for="test-1"
  >
    <input
      class="toggle-input__input"
      id="test-1"
      name="test"
      type="radio"
      value="foo"
    />
     
    <div
      class="radio-indicator radio-indicator--normal radio-indicator--standard radio-indicator--checked"
    />
     
    <div
      class="toggle-input__label"
    >
      <span>
        Option 2
      </span>
    </div>
  </label>
  <label
    class="toggle-input radio-input toggle-input--normal toggle-input--standard toggle-input--label-right"
    for="test-2"
  >
    <input
      class="toggle-input__input"
      id="test-2"
      name="test"
      type="radio"
      value="bar"
    />
     
    <div
      class="radio-indicator radio-indicator--normal radio-indicator--standard"
    />
     
    <div
      class="toggle-input__label"
    >
      <span>
        Option 3
      </span>
    </div>
  </label>
</div>
`;
